pip-lock/main.ts

```diff
@@ -0,0 +1,392 @@
+    if(testPath) {
+        const path = await tryRunPip(testPath);
+        if(path)
+            return path;
+    }
+    const firstTry = await tryRunPip("pip3");
+    if(!firstTry)
+        return await tryRunPip("pip");
+
+    return firstTry;
+}
+
+// Provides back the parsed version of the current pip installation.
+async function getVersion(pipPath: string): Promise<PipVersionTuple> {
+    const res = await Deno.run({
+        cmd: [pipPath, '--version'],
+        stdout: 'piped',
+        stderr: 'piped'
+    });
+
+    try {
+        const data = await res.output();
+
+        const textResult = new TextDecoder().decode(data);
+
+        const matchData = textResult.match(/pip\s+(\d+\.\d+\.\d+).*/);
+
+        if (!matchData || matchData.length !== 2)
+            throw new Error("unexpected output from pip");
+        // This should be the pip version
+        const versionData = matchData[1];
+        const versionOutput = versionData.split('.');
+        if(versionOutput.length < 3)
+            throw new Error("unexpected version information");
+
+        return {
+            major: parseInt(versionOutput[0]),
+            minor: parseInt(versionOutput[1]),
+            fix: parseInt(versionOutput[2]),
+        };
+    } finally {
+        res.close();
+    }
+}
+
+
+
+const binaryFormats = new Set(['.tar.gz', '.zip', '.whl', '.pyd']);
+
+// Extracts the package version from the target string, which is of the form:
+// Flask-2.2.2-py3-none-any.whl
+// This is made complicated due to the fact that package names may not match
+// the provided packageName, and can contain -, _, ., etc. Versions may not
+// adhere to pep8 standards also, so that must be accounted for here.
+// This method will ultimately return the package/version tuple.
+function parsePackageVersion(packageName: string, target: string): string[] {
+    if(target.length < (packageName.length+2))
+        throw new Error(`invalid target ${target} does not contain the package name ${packageName}`);
+
+    // We will parse the real package name out of the target string
+    // if it isn't the same as the package name provided. This happens with some packages -
+    // consider flask shows up as Flask here. At present, our heuristic is that the beginning of the
+    // next field (which is the package version) must start with the pattern `-[0-9]`; there are no
+    // other guarantees, so that is what we start by looking for.
+    let fixedName = packageName;
+    if(!target.startsWith(packageName)) {
+        const tmp: string[] = [];
+        for(let i = 0; i < target.length; i++) {
+            if(target[i] === '-' && i < (target.length - 1)) {
+                const check = target[i+1];
+                // If value is ascii '0'-'9'
+                if(check.charCodeAt(0) <= 57 && check.charCodeAt(0) >= 48)
+                    break;
+            }
+
+            tmp.push(target[i]);
+        }
+
+        fixedName = tmp.join('');
```

Comment:
Can all of this be replaced by `fixedName = target.replace(/-\d.*/, '')`?

Reply:
We are just parsing things from messy output, not trying to clean them up, so replace is probably not the right choice here.

Reply:
Replace returns a new string containing only the part that's being parsed out here. Alternatively, you could write something like `fixedName = (target.match(/.*?(?=-\d)/) || [target])[0];`, but it seems much less clear.